In [57]:
import pandas as pd

orders_df = pd.read_csv("C:/Users/Kelley/Desktop/Project/Final Project/data/olist_orders_dataset.csv")

orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [58]:
# Exploring the orders dataset.
orders_df.info()
orders_df.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

### count of order_id = unique of order_id, meaning there is no duplicate orders

### based on above, must change data type

 0   order_id                       99441 non-null  object  No change  

 1   customer_id                    99441 non-null  object  No change
 
 2   order_status                   99441 non-null  object  to  category    also a cat table?
 
 3   order_purchase_timestamp       99441 non-null  object  to  timedate
 
 4   order_approved_at              99281 non-null  object  to  timedate
 
 5   order_delivered_carrier_date   97658 non-null  object  to  timedate
 
 6   order_delivered_customer_date  96476 non-null  object  to  timedate
 
 7   order_estimated_delivery_date  99441 non-null  object  to  timedate

In [59]:
# Converting order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date
orders_df['order_purchase_timestamp'] = pd.to_datetime(orders_df['order_purchase_timestamp'])
orders_df['order_approved_at'] = pd.to_datetime(orders_df['order_approved_at'])
orders_df['order_delivered_carrier_date'] = pd.to_datetime(orders_df['order_delivered_carrier_date'])
orders_df['order_delivered_customer_date'] = pd.to_datetime(orders_df['order_delivered_customer_date'])
orders_df['order_estimated_delivery_date'] = pd.to_datetime(orders_df['order_estimated_delivery_date'])

orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


In [60]:
# Returning unique values for order status.
orders_df["order_status"].unique()


array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

In [61]:
# Replaced wrong spelling 'canceled' to 'cancelled'
orders_df["order_status"] = orders_df["order_status"].replace("canceled", "cancelled")

In [62]:
# Returning unique values for order status in a dataframe.
order_status_unique = orders_df[["order_status"]].drop_duplicates(ignore_index=True)

# Reseting index to start from 1.
order_status_unique.index = order_status_unique.index + 1

print(order_status_unique)

# Checking and confirming type of order_status_unique to be a dataframe.
if isinstance(order_status_unique, pd.DataFrame):
    print("order_status_unique is a DataFrame.")

  order_status
1    delivered
2     invoiced
3      shipped
4   processing
5  unavailable
6    cancelled
7      created
8     approved
order_status_unique is a DataFrame.


### Exploring relationship of null values Vs order_status

 0   order_id                       99441 non-null  object  

 1   customer_id                    99441 non-null  object
 
 2   order_status                   99441 non-null  object
 
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]

 4   order_approved_at              99281 non-null  datetime64[ns]
 
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]

### Exploring nulls in 'order_approved_at'

In [63]:
# Subsetting with order_approved_at nulls
approved_null = orders_df[orders_df['order_approved_at'].isnull()]

approved_null.head()

order_id                       customer_id  \
1130  00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801  ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868  df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029  8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161  7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   

     order_status order_purchase_timestamp order_approved_at  \
1130    cancelled      2018-08-28 15:26:39               NaT   
1801    cancelled      2018-09-20 13:54:16               NaT   
1868    cancelled      2017-03-04 12:14:30               NaT   
2029    cancelled      2018-08-29 16:27:49               NaT   
2161    cancelled      2017-05-01 16:12:39               NaT   

     order_delivered_carrier_date order_delivered_customer_date  \
1130                          NaT                           NaT   
1801                          NaT                           NaT   
1868                          NaT                           NaT   
2029                          NaT                           NaT   
2161                          NaT                           NaT   

     order_estimated_delivery_date  
1130                    2018-09-12  
1801                    2018-10-17  
1868                    2017-04-10  
2029                    2018-09-13  
2161                    2017-05-30

In [64]:
unique_row_count_approved_null = approved_null.nunique()

print(unique_row_count)

order_id                         160
customer_id                      160
order_status                       3
order_purchase_timestamp         160
order_approved_at                  0
order_delivered_carrier_date      13
order_delivered_customer_date     14
order_estimated_delivery_date    109
dtype: int64


In [65]:
# Returning unique values for order status for approved_null df
approved_null["order_status"].unique()

array(['cancelled', 'delivered', 'created'], dtype=object)

In [66]:
# Value count of order status
value_counts = approved_null["order_status"].value_counts()
print(value_counts)


order_status
cancelled    141
delivered     14
created        5
Name: count, dtype: int64


#### At time of data collection, 
#### 141 orders were cancelled, 5 orders were created and therefore had null values for order_approved_at
#### 14 orders were delivered even though order_approved_at was null - to investigate further if required

### Exploring nulls in order_delivered_customer_date

In [67]:
# Subsetting with order_delivered_customer_date nulls
delivered_customer_null = orders_df[orders_df['order_delivered_customer_date'].isnull()]

delivered_customer_null.head()

order_id                       customer_id  \
6    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp   order_approved_at  \
6       invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44       shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103     invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128   processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154      shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
6                            NaT                           NaT   
44           2018-06-05 14:32:00                           NaT   
103                          NaT                           NaT   
128                          NaT                           NaT   
154          2018-01-11 19:39:23                           NaT   

    order_estimated_delivery_date  
6                      2017-05-09  
44                     2018-06-28  
103                    2018-08-21  
128                    2017-10-03  
154                    2018-02-07

In [70]:
unique_row_count_delivered_customer_null = delivered_customer_null.nunique()

print(unique_row_count_delivered_customer_null)

order_id                         2965
customer_id                      2965
order_status                        8
order_purchase_timestamp         2962
order_approved_at                2806
order_delivered_carrier_date     1178
order_delivered_customer_date       0
order_estimated_delivery_date     424
dtype: int64


In [ ]:
# Returning unique values for order status for deleivered_customer_null df
delivered_customer_null["order_status"].unique()

array(['invoiced', 'shipped', 'processing', 'unavailable', 'cancelled',
       'delivered', 'created', 'approved'], dtype=object)

In [ ]:
# Value count of order status
value_counts = delivered_customer_null["order_status"].value_counts()
print(value_counts)


order_status
shipped        1107
cancelled       619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64


### 8 orders have status=delivered but delivered_customer_date is null - to investigate further if required or Drop

### checking delivered_customer_date with order_status

In [ ]:
# 
# Assuming you have a DataFrame named orders_df
deliveredcust = orders_df.dropna(subset=["order_delivered_customer_date"]).copy()



In [ ]:
# Returning unique values for order status for deleivered_customer_null df
deliveredcust["order_status"].unique()

array(['delivered', 'cancelled'], dtype=object)

In [ ]:
# Value count of order status
value_counts = deliveredcust["order_status"].value_counts()
print(value_counts)


order_status
delivered    96470
cancelled        6
Name: count, dtype: int64


### there are 6 cancelled orders but have delivered to customer timestamp - to investigate if required or drop